src/v/cluster/cluster_recovery_table.h

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include "cluster/cloud_metadata/cluster_manifest.h"
+#include "cluster/cluster_recovery_state.h"
+#include "cluster/controller_snapshot.h"
+#include "cluster/types.h"
+#include "model/fundamental.h"
+
+namespace cluster {
+
+// Returns true if the given status implies there may be more state to be
+// recovered from a remote controller snapshot.
+inline bool
+may_require_controller_recovery(std::optional<recovery_stage> cur_status) {
+    if (!cur_status.has_value()) {
+        return true;
+    }
+    switch (cur_status.value()) {
+    case recovery_stage::recovered_controller_snapshot:
+    case recovery_stage::recovered_offsets_topic:
+    case recovery_stage::recovered_tx_coordinator:
+    case recovery_stage::complete:
+    case recovery_stage::failed:
+        return false;
+    default:
+        return true;
+    }
+};
+
+// Tracks the state of recovery attempts performed on the cluster.
+//
+// It is expected this will live on every shard on every node, making it easy
+// to determine from any shard what the current status of recovery is (e.g. for
+// reporting, or to install a guardrail while recovery is running).
+class cluster_recovery_table {
+public:
+    ~cluster_recovery_table() { _has_active_recovery.broken(); }
```

Comment:
more appropriate for a `stop()` interface?

Reply:
Done